dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemetryInfoFactoryImpl.java

```diff
@@ -0,0 +1,331 @@
+
+        OpenTelemetryInfoWrappedSupplier newSupplier = new OpenTelemetryInfoWrappedSupplier(openTelemetryInfo -> openTelemetryInfo.dispose(), this::createOpenTelemetryInfo);
+
+        if (oTelRef == null) {
+            oTelRef = new OpenTelemetryInfoReference();
+        }
+        oTelRef.set(newSupplier);
+        extAppInfo.getMetaData().setMetaData(slotForOpenTelemetryInfoHolder, oTelRef);
+    }
+
+    @Override
+    public void applicationStarted(ApplicationInfo appInfo) throws StateChangeException {} //no-op
+
+    @Override
+    public void applicationStopping(ApplicationInfo appInfo) {} //no-op
+
+    @Override
+    public void applicationStopped(ApplicationInfo appInfo) {
+        ExtendedApplicationInfo extAppInfo = (ExtendedApplicationInfo) appInfo;
+        OpenTelemetryInfoReference oTelRef = (OpenTelemetryInfoReference) extAppInfo.getMetaData().getMetaData(slotForOpenTelemetryInfoHolder);
+
+        OpenTelemetryInfoWrappedSupplier newSupplier = new OpenTelemetryInfoWrappedSupplier(openTelemetryInfo -> {
+        }, OpenTelemetryInfoFactoryImpl::createDisposedOpenTelemetryInfo);
+
+        OpenTelemetryInfoWrappedSupplier oldSupplier = oTelRef.getAndSet(newSupplier);
+
+        try {
+            oldSupplier.closeAndDisposeIfCreated();
+        } catch (Exception e) {
+            Tr.warning(tc, "applicationStopped", "failed to dispose of OpenTelemetry");//TODO better message
+        }
+
+    }
+
+    //Adds the service name to the resource attributes
+    private static Resource customizeResource(Resource resource, ConfigProperties c) {
+        ResourceBuilder builder = resource.toBuilder();
+        builder.put(ResourceAttributes.SERVICE_NAME, getServiceName(c));
+        return builder.build();
+    }
+
+    //Uses application name if the user has not given configured service.name resource attribute
+    private static String getServiceName(ConfigProperties c) {
+        String appName = c.getString(SERVICE_NAME_PROPERTY);
+        ComponentMetaData cmd = ComponentMetaDataAccessorImpl.getComponentMetaDataAccessor().getComponentMetaData();
+        if (appName == null) {
+            if (cmd != null) {
+                appName = cmd.getModuleMetaData().getApplicationMetaData().getName();
+            }
+        }
+        return appName;
+    }
+
+    //Interfaces and private classes only relevent to this factory.
+
+    public interface OpenTelemetrySdkBuilderSupplier {
+        public AutoConfiguredOpenTelemetrySdkBuilder getPartiallyConfiguredOpenTelemetrySDKBuilder();
+    }
+
+    /*
+     * There are two race conditions we need to protect against.
+     *
+     * On app startup/shutdown we need to swap the holder associated with the application.
+     * We protect against race conditions here by using an AtomicReference.
+     *
+     * Within the context of an application's lifecycle we need to ensure OpenTelemetryInfo
+     * is only created once. OpenTelemetryInfoReference handles this.
+     */
```

Comment:
Looking at this again, it's not clear to me why we need to swap the holder associated with the application. 

We've ended up now with this chain of objects:
App Metadata -> OpenTelemetryInfoReference -> OpenTelemetryInfoWrappedSupplier -> OpenTelemetryInfo -> OpenTelemetry

Of these, I can explain the jobs of two:
- `OpenTelemetryInfoWrappedSupplier`: responsible for lazily initializing the `OpenTelemetryInfo` and disposing of it when the app shuts down.
- `OpenTelemetryInfo`: stores the `OpenTelemetry` and `Tracer` objects for when they need to be looked up. Also stores whether the `OpenTelemetry` is a real implementation or a no-op so that some integrations can be disabled if OTel is disabled.

I'm not sure why `OpenTelemetryInfoReference` is needed. At the moment we only update this when the app shuts down so that it will hold a different `OpenTelemetryInfoWrappedSupplier` which will create and return a `DisposedOpenTelemetryInfo` if asked.

I'm not sure why we couldn't instead update the `OpenTelemetryInfo` instance that `OpenTelemetryInfoWrappedSupplier` is holding when it disposes the old one.

Reply:
I guess you could say that the job of the `OpenTelemetryInfoWrappedSupplier` is just to handle lazily initializing and destroying the `OpenTelemetryInfo` - and then we need something else if we want to actually change it for a different object?

It does feel a bit like we're doing half the job in one place and half in another which makes it tricky to see what's going on.

Will think about it more tomorrow.

Reply:
It would work if we updated the instance within OpenTelemetryInfoWrappedSupplier. 

However I think its a little neater this way. It gives OpenTelemetryInfoWrappedSupplier a clear lifecycle and takes fewer LOC.

Reply:
ok